code/modules/projectiles/guns/energy/hos.dm

```diff
@@ -0,0 +1,304 @@
+
+	fire_sound = 'sound/weapons/Taser.ogg'
+
+	mag_type = /obj/item/ammo_box/magazine/mws_mag
+	spawnwithmagazine = FALSE
+
+	recoil = 0
+
+	var/charge_sections = 6
+
+/obj/item/gun/ballistic/revolver/mws/afterattack(atom/target, mob/living/user, flag, params)
+	if(chambered && !chambered.BB) //if BB is null, i.e the shot has been fired...
+		var/obj/item/ammo_casing/mws_batt/shot = chambered
+		if(shot.shots_left >= 1)
+			shot.newshot() //Create a new shot
+			update_charge()
+	.=..()
+
+/obj/item/gun/ballistic/revolver/mws/process_chamber()
+	if(chambered && !chambered.BB) //if BB is null, i.e the shot has been fired...
+		var/obj/item/ammo_casing/mws_batt/shot = chambered
+		if(shot.shots_left >= 1)
+			shot.newshot() //Create a new shot
+			update_charge()
+		else
+			for(var/B in magazine.stored_ammo)
+				var/obj/item/ammo_casing/mws_batt/other_batt = B
+				if(istype(other_batt,shot) && other_batt.shots_left >= 1)
+					switch_to(other_batt)
+	else
+		return
+
+/obj/item/gun/ballistic/revolver/mws/proc/update_charge()
+
+	if(!chambered)
+		return
+
+	var/obj/item/ammo_casing/mws_batt/batt = chambered
+	if(batt.cell.charge >= batt.e_cost)
+		batt.cell.use(batt.e_cost)
+
+	if(batt.cell.charge > 0)
+		batt.shots_left = (batt.cell.charge / batt.cell.maxcharge) * 7
+	else
+		batt.shots_left = 0
+	update_icon()
+
+/obj/item/gun/ballistic/revolver/mws/proc/switch_to(obj/item/ammo_casing/mws_batt/new_batt)
+	if(ishuman(loc))
+		if(chambered && new_batt.type == chambered.type)
+			to_chat(loc,"<span class='warning'>\The [src] is now using the next [new_batt.type_name] power cell.</span>")
+		else
+			to_chat(loc,"<span class='warning'>\The [src] is now firing [new_batt.type_name].</span>")
+
+	chambered = new_batt
+	update_icon()
+
+/obj/item/gun/ballistic/revolver/mws/attack_self(mob/living/user)
+	if(!chambered)
+		return
+
+	var/list/stored_ammo = magazine.stored_ammo
+
+	if(stored_ammo.len == 1)
+		return //silly you.
+
+	//Find an ammotype that ISN'T the same, or exhaust the list and don't change.
+	var/our_slot = stored_ammo.Find(chambered)
+
+	for(var/index in 1 to stored_ammo.len)
+		var/true_index = ((our_slot + index - 1) % stored_ammo.len) + 1 // Stupid ONE BASED lists!
+		var/obj/item/ammo_casing/mws_batt/next_batt = stored_ammo[true_index]
+		if(chambered != next_batt && !istype(next_batt, chambered.type) && next_batt.shots_left >= 1)
+			switch_to(next_batt)
+			break
+
+/obj/item/gun/ballistic/revolver/mws/AltClick(mob/living/user)
+	if(magazine)
+		magazine.forceMove(drop_location())
+		user.put_in_hands(magazine)
+		magazine.update_icon()
+		if(magazine.ammo_count())
+			playsound(src, 'sound/weapons/gun_magazine_remove_full.ogg', 70, 1)
+		else
+			playsound(src, "gun_remove_empty_magazine", 70, 1)
+		magazine = null
+		to_chat(user, "<span class='notice'>You pull the magazine out of \the [src].</span>")
+		if(chambered)
+			chambered = null
+		update_icon()
+
+/obj/item/gun/ballistic/revolver/mws/update_icon()
+
```

Comment:
everything in here dealing with overlays needs to be update_icon_overlays() instead
plus pretty sure it's better to use mutable_appearances here instead